File: datastruct/queue/LoopQueue.java
import java.util.Objects;

public class LoopQueue<E> implements Queue<E> {

    private E[] data;
    private int front; // 队首元素位置
    private int tail;  // 下一个被传入元素的位置
    private int size;  // 队列内元素个数

    public LoopQueue(int capacity){
        data = (E[])new Object[capacity + 1];  // 循环队列会浪费一个空间
    }

    public LoopQueue(){
        this(10);
    }

    @Override
    public boolean isEmpty(){
        return front == tail;
    }

    @Override
    public int getSize(){
        return size;
    }

    public int getCapacity(){
        return  data.length - 1;
    }

    @Override
    public void enqueue(E ele){

        if((tail + 1) % data.length == front)
            resize(getCapacity() * 2);

        data[tail] = ele;
        tail = (tail + 1) % data.length;
        size++;
    }

    @Override
    public E dequeue(){
        if(front == tail)
            throw new IllegalArgumentException("Cannot dequeue from empty queue!");

        E res = data[front];
        data[front] = null;

        front = (front + 1) % data.length;
        size--;
        if(size < getCapacity() / 4)
            resize(getCapacity() / 2);
        return res;

    }

    @Override
    public E getFront(){
        return data[front];
    }

    private void resize(int newCapacity){
        E[] newdata = (E[])new Object[newCapacity + 1];
        for(int i = 0; i < size; i++){
            newdata[i] = data[(i + front) % data.length];
        }
        front = 0;
        tail = size;
        data = newdata;
    }

    @Override
    public String toString(){
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("LoopQueue size: %d, capacity: %d\n", size, getCapacity()));
        sb.append("front [");
        for(int i = front; i != tail; i = (i + 1) % data.length){
            sb.append(data[i]);

            if((i + 1) % data.length != tail)
                sb.append(", ");
        }
        sb.append("] tail");
        return  sb.toString();
    }
}

File: datastruct/queue/LoopQueue.py
from Queue import Queue


class LoopQueue(Queue):

    def __init__(self, capacity: int = 10):

        self.__data = [None] * (capacity + 1)
        self.__front = 0
        self.__tail = 0
        self.__size = 0

    def get_size(self):
        return self.__size

    def is_empty(self):
        return self.__front == self.__tail

    def enqueue(self, ele):

        if (self.__tail + 1) % len(self.__data) == self.__front:
            self.__resize((len(self.__data) - 1) * 2)

        self.__data[self.__tail] = ele
        self.__tail = (self.__tail + 1) % len(self.__data)
        self.__size += 1

    def __resize(self, new_capacity: int):

        new_data = [None] * (new_capacity + 1)

        for i in range(self.__size):
            new_data[i] = self.__data[(i + self.__front) % len(self.__data)]
        self.__data = new_data
        self.__front = 0
        self.__tail = self.__size

    def dequeue(self):

        if self.__size == 0:
            raise IndexError('Cannot dequeue from an empty queue!')


        res = self.__data[self.__front]
        self.__data[self.__front] = None
        self.__front = (self.__front + 1) % (len(self.__data))
        self.__size -= 1

        if self.__size < self.get_capacity() / 4:
            self.__resize(self.get_capacity() // 2)

        return res

    def get_front(self):
        return self.__data[self.__front]

    def __str__(self):

        bstr = 'LoopQueue size: %d, capacity: %d\n' % (self.__size, len(self.__data))
        bstr += 'front ['
        i = self.__front
        while i != self.__tail:
            bstr += '' + str(self.__data[i])

            if (i + 1) % len(self.__data) != self.__tail:
                bstr += ', '

            i = (i + 1) % len(self.__data)
        bstr += '] tail'
        return bstr

    def get_capacity(self):
        return len(self.__data) - 1
